### ISSUE TITLE
Should we maintain 2 separate read_hapmap functions, and a single write_hapmap?

### ISSUE THREAD

**Issue by hyanwong** [tskit-dev/msprime#1338]
Opening this to seek consensus between @jeromekelleher and @grahamgower. At the moment we have 2 functions:  `msprime.RecombinationMap.read_hapmap()` (deprecated but maintained) and `msprime.read_hapmap()`. Currently the first is a thin wrapper for the second. As far as I understand, @grahamgower suggested separating out the logic so that `msprime.RecombinationMap.read_hapmap()` carried on reading the "rate" column (not the recommended data source), whereas `msprime.read_hapmap()` would only read the "genetic position (cM)" position (the canonical data source). Additionally `msprime.read_hapmap()` would be more strict about the data format required. However, I think @jeromekelleher wasn't convinced by the code duplication that would entail.

As part of this, @grahamgower suggested that an `msprime.RecombinationMap.write_hapmap()` function would be useful to output the lower quality maps in `stdpopsim` to a format that would then be readable by `msprime.read_hapmap()`. @jeromekelleher thought that perhaps this would be useful for RateMaps in general, and so should be a method of the general RateMap class. However, a RateMap is more general than a recombination map, so the headers that would be output would probably not be suitable for general use. I guess we could have a `RateMap.write_text(name=XXX, headers=XXX)` function which required the user to provide headers - `RecombinationMap.write_hapmap()` could then call this with the appropriate headers.

Before coding this up, I'd like some clarity on what the approach should be - thanks!

**Comment by jeromekelleher** [comment]
Thanks for this @hyanwong, good summary.

Ideally, I'd like to see the following:

1. An implementation RateMap.read_hapmap that can read from *either* the rate or map column, which looks like 

```
def read_hapmap(fileobj, position_col=None, rate_col=None, map_col=None):
     position_col = 1 if position_col is None else position_col
     if rate_col is None and map_col is None:
          # Default to map_col
          map_col = 3
    elif rate_col is not None and map_col is not None:
         raise ValueError("Cannot specify both rate_col and map_col")
   if map_col is not None:
         # read in the hapmap from the map col
   else:
         assert rate_col is not None
         # read in the hapmap from map_col
```

Then, RecombinationMap.read_hapmap() uses this implementation to continue supporting the existing functionality.

We can add RateMap.write_hapmap(). This would just use the format - if people happen to use this for non recombination map data, then so be it. If we want a more general write_text method we can add it later.

**Comment by grahamgower** [comment]
Looks good to me.

**Comment by hyanwong** [comment]
OK, thanks both. That should be enough for me to be getting on with.

**Comment by jeromekelleher** [comment]
closed in #1346.
